File: Algorithms on Strings/Assignment1/trie.py
#Uses python3
import sys

# Return the trie built from patterns
# in the form of a dictionary of dictionaries,
# e.g. {0:{'A':1,'T':2},1:{'C':3}}
# where the key of the external dictionary is
# the node ID (integer), and the internal dictionary
# contains all the trie edges outgoing from the corresponding
# node, and the keys are the letters on those edges, and the
# values are the node IDs to which these edges lead.
def build_trie(patterns):
    tree = dict()
    tree[0] = {}
    node_count = 1
    for pattern in patterns:
        curr_node = 0
        for c in pattern:
            if c not in tree[curr_node].keys():
                tree[curr_node][c] = node_count
                tree[node_count] = {}
                node_count += 1
            curr_node = tree[curr_node][c]
    return tree


if __name__ == '__main__':
    patterns = sys.stdin.read().split()[1:]
    #patterns = ['ATAGA', 'ATC', 'GAT']
    tree = build_trie(patterns)
    for node in tree:
        for c in tree[node]:
            print("{}->{}:{}".format(node, tree[node][c], c))

File: Algorithms on Strings/Assignment1/trie.cpp
#include <string>
#include <iostream>
#include <vector>
#include <map>

using namespace std;

typedef map<char, int> edges;
typedef vector<edges> trie;

bool find_val(map<char, int> edges, char key) {
    map<char, int>::iterator it;
    it = edges.find(key);
    if (it != edges.end()) {
        return true;
    }
    return false;
}

trie build_trie(vector<string> & patterns) {
    trie t;
    t.push_back(map<char, int>());
    int node_count = 1;
    for (string pattern: patterns) {
        int curr_node = 0;
        for (char c: pattern) {
            if (!find_val(t[curr_node], c)) {
                t[curr_node].insert(make_pair(c, node_count));
                t.push_back(map<char, int>());
                node_count++;
            }
            curr_node = t[curr_node][c];
        }
    }
    return t;
}

int main() {
    size_t n;
    std::cin >> n;
    vector<string> patterns;
    for (size_t i = 0; i < n; i++) {
        string s;
        std::cin >> s;
        patterns.push_back(s);
    }
    
    trie t = build_trie(patterns);
    for (size_t i = 0; i < t.size(); ++i) {
        for (const auto & j : t[i]) {
            std::cout << i << "->" << j.second << ":" << j.first << "\n";
        }
    }
    
    return 0;
}
